spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,301 @@
+    return StreamingOffset.fromJson(json);
+  }
+
+  @Override
+  public void commit(Offset end) {
+    // Since all the data and metadata of Iceberg is as it is, nothing needs to commit when
+    // offset is processed, so no need to implement this method.
+  }
+
+  @Override
+  public void stop() {}
+
+  @Override
+  @SuppressWarnings("unchecked")
+  protected List<CombinedScanTask> tasks() {
+    if (startOffset.equals(endOffset)) {
+      LOG.info("Start offset {} equals to end offset {}, no data to process", startOffset, endOffset);
+      return Collections.emptyList();
+    }
+
+    Preconditions.checkState(pendingTasks != null,
+        "pendingTasks is null, which is unexpected as it will be set when calculating end offset");
+
+    if (Iterables.isEmpty(pendingTasks)) {
+      LOG.info("There's no task to process in this batch");
+      return Collections.emptyList();
+    }
+
+    IndexedTask last = Iterables.getLast(pendingTasks);
+    if (last.snapshotId() != endOffset.snapshotId() || last.index() != endOffset.index()) {
+      throw new IllegalStateException("The cached pendingTasks doesn't match the current end offset " + endOffset);
+    }
+
+    LOG.info("Processing data from {} exclusive to {} inclusive", startOffset, endOffset);
+    CloseableIterable<FileScanTask> splitTasks = TableScanUtil.splitFiles(
+        CloseableIterable.transform(pendingTasks, IndexedTask::task), splitSize());
+    return Lists.newArrayList(
+        TableScanUtil.planTasks(splitTasks, splitSize(), splitLookback(), splitOpenFileCost()));
+  }
+
+  private StreamingOffset calculateStartingOffset() {
+    StreamingOffset startingOffset;
+    if (startSnapshotId != null) {
+      startingOffset = new StreamingOffset(startSnapshotId, -1, true);
+    } else {
+      List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+      if (snapshotIds.isEmpty()) {
+        // there's no snapshot currently.
+        startingOffset = StreamingOffset.START_OFFSET;
+      } else {
+        startingOffset = new StreamingOffset(snapshotIds.get(snapshotIds.size() - 1), -1, true);
+      }
+    }
+
+    return startingOffset;
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    if (start.equals(StreamingOffset.START_OFFSET)) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    this.pendingTasks = getChangesWithRateLimit(start.snapshotId(), start.index(),
+        start.isStartingSnapshotId(), maxFilesPerBatch);
+    IndexedTask last = Iterables.isEmpty(pendingTasks) ? null : Iterables.getLast(pendingTasks);
+
+    if (last == null) {
+      return start;
+    } else {
+      boolean isStarting = last.snapshotId() == start.snapshotId() && start.isStartingSnapshotId();
+      return new StreamingOffset(last.snapshotId(), last.index(), isStarting);
+    }
+  }
+
+  @VisibleForTesting
+  CloseableIterable<IndexedTask> getChangesWithRateLimit(long snapshotId, int index, boolean isStarting, int maxFiles) {
```

Comment:
What about defining the size of each micro-batch in terms of data size and not in terms of the number of files or number of tasks? I think the whole idea of rate-limiting is not to process too much data in one batch. The data size seems more stable than the number of files and the number of tasks should depend on the data size and other table configs.